GameServer/gameobjects/GameMerchant.cs

```diff
@@ -81,121 +82,132 @@ public virtual void SendMerchantWindow(GamePlayer player)
+                {
+                    var currencyItem = page.CurrencyItem;
+                    var costToText = CurrencyToText(currency.Mint(cost));
+                    if (!HasPlayerEnoughBalance(player, currency.Mint(cost)))
+                    {
+                        player.Out.SendMessage(LanguageMgr.GetTranslation(player.Client.Account.Language, "GameMerchant.OnPlayerBuy.YouNeedGeneric", costToText), eChatType.CT_System, eChatLoc.CL_SystemWindow);
+                        return;
+                    }
+                    if (!player.Inventory.AddTemplate(GameInventoryItem.Create(itemToBuy), amountToBuy, eInventorySlot.FirstBackpack, eInventorySlot.LastBackpack))
+                    {
+                        player.Out.SendMessage(LanguageMgr.GetTranslation(player.Client.Account.Language, "GameMerchant.OnPlayerBuy.NotInventorySpace"), eChatType.CT_System, eChatLoc.CL_SystemWindow);
+                        return;
+                    }
+                    InventoryLogging.LogInventoryAction(this, player, eInventoryActionType.Merchant, itemToBuy, amountToBuy);
+
+                    string message;
+                    if (amountToBuy == 1)
+                        message = LanguageMgr.GetTranslation(player.Client.Account.Language, "GameMerchant.OnPlayerBuy.BoughtGeneric", itemToBuy.Name, costToText);
+                    else
+                        message = LanguageMgr.GetTranslation(player.Client.Account.Language, "GameMerchant.OnPlayerBuy.BoughtPiecesGeneric", amountToBuy, itemToBuy.Name, costToText);
+                    WithdrawCurrencyFromPlayer(player, currency.Mint(cost));
+                    player.Out.SendMessage(message, eChatType.CT_Merchant, eChatLoc.CL_SystemWindow);
+
+                }
+                else if (currency.Equals(Currency.Mithril)) throw new NotImplementedException("Mithril is currently not implemented as a separate currency.");
+                else throw new ArgumentException($"{currency} is not implemented.");
+            }
+        }
+
+        private void WithdrawCurrencyFromPlayer(GamePlayer player, Money price)
+        {
+            if (price.Currency.Equals(Currency.Copper))
+            {
+                if (!player.RemoveMoney(price.Amount)) throw new Exception("Money amount changed while adding items.");
+            }
+            else if (price.Currency.Equals(Currency.BountyPoints)) player.BountyPoints -= price.Amount;
+            else if (price.Currency is ItemCurrency itemCurrency)
+            {
+                player.Inventory.RemoveTemplate(itemCurrency.Item.Id_nb, (int)price.Amount, eInventorySlot.FirstBackpack, eInventorySlot.LastBackpack);
+            }
+            else throw new NotImplementedException($"{price.Currency} is currently not implemented.");
+        }
+
+        private bool HasPlayerEnoughBalance(GamePlayer player, Money price)

```

Comment:
Currency is not responsible for money balance. Money is neither btw.

Reply:
I think it can be good to have another class (maybe just a static class) to manage the money balance, doing it in the Currency directly make it easier for the dispatch.
I don't know if it's possible to do it in another way in C# without managing the dispatch manually.

Maybe I need to explain why I'm talking about that: your changes are very good, I really like what you did, I just see an opportunity to make it more extensible. With the current code, you can't easily add a new currency but I think with very few changes it can be possible to make it happen and I can see a good use for custom servers (like using the XP for paying something, maybe your toon life, etc).

Reply:
This is likely going to be addressed in the future, however not in this PR. For now it was just to divide functionality and implicitly flag polymorphism (verbose use of if/else). It should increase readability, but also make the PR at least somewhat focussed.